Add `tracked-built-ins` dependency (#812)
emberjs/rfcs: text/0812-tracked-built-ins.md

---
Stage: Accepted
Start Date: 2022-03-29
Release Date: Unreleased
Release Versions:
  ember-source: vX.Y.Z
  ember-data: vX.Y.Z
Relevant Team(s): Ember CLI, Learning
RFC PR: https://github.com/emberjs/rfcs/pull/812
---

<!--- 
Directions for above: 

Stage: Leave as is
Start Date: Fill in with today's date, YYYY-MM-DD
Release Date: Leave as is
Release Versions: Leave as is
Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
RFC PR: Fill this in with the URL for the Proposal RFC PR
-->

# Add tracked-built-ins

## Summary

This RFC proposes adding [tracked-built-ins](https://github.com/tracked-tools/tracked-built-ins)

> to the blueprints that back `ember new` and `ember addon`.

Reviewer: Whoops, please ignore this—this was from a long-outdated conversation. That *is* probably what we will do long-term, but if I recall correctly we wanted to do that after further iteration. I'll make sure we discuss it before putting it into FCP!

that allow to implement tracked versions of JavaScript's built-in classes
in an addon like [tracked-built-ins](https://github.com/tracked-tools/tracked-built-ins):

```ts
import {
  TrackedObject,
  TrackedArray,
  TrackedMap,
  TrackedSet,
  TrackedWeakMap,
  TrackedWeakSet,
} from 'tracked-built-ins';

class Foo {
  @tracked value = 123;

  obj = new TrackedObject<{[string: number]}>({ foo: 123 });
  arr = new TrackedArray<string, number>(['foo', 123]);
  map = new TrackedMap<string, number>(['foo', 123]);
  set = new TrackedSet<string, number>(['foo', 123]);
  weakMap = new TrackedWeakMap<{[object: number]}>([[{}, 456]]);
  weakSet = new TrackedWeakSet<string, number>(['foo', 123]);
}
```

or using `tracked()` decorator:

```ts
import { tracked } from 'tracked-built-ins';

class Foo {
  @tracked value = 123;

  obj = tracked<{[string: number]}>({ foo: 123 });
  arr = tracked<string, number>(['foo', 123]);
  map = tracked<string, number>(new Map(['foo', 123]));
  set = tracked<string, number>(new Set(['foo', 123]));
  weakMap = tracked<{[string: number]}>(new WeakMap([[{}, 456]]));
  weakSet = tracked<string, number>(new WeakSet(['foo', 123]));
}

> ```

Reviewer: Raising this explicitly for discussion: I do not think we should include this part of the `tracked-built-ins` API.

Reviewer 2: I agree with this. In practice, I've found that importing the decorator from `tracked-built-ins` gets confusing. In some components, you'll import it from `@glimmer/tracking`, and in others you might use `tracked-built-ins` instead. I find it much clearer to import the classes and instantiate them instead. We've even found rare cases where we need to use both:

```
@tracked someMap = new TrackedMap();
```

Being able to consistently understand what `@tracked` is capable of no matter which file you're looking at without having to check where it's imported from is much better for developer ergonomics, IMO.

Author: I tend to agree with this - `tracked()` function may be *misleading* and you need to be careful what you use and where.

would be puzzle to have in the file both `tracked` provided by `@glimmer/tracking` *and* `tracked` provided by `tracked-built-ins` as you would have to rename one of those in the import statement

## Detailed design

The necessary changes to `ember-cli` are relatively small since we only need
to add the dependency to the `app` blueprint, and the `addon` blueprint will
inherit it automatically.

This has the advantage (over including it as an implicit dependency), that
apps and addons that don't want to use it for some reason can opt out by
removing the dependency from their `package.json` file.

## How we teach this

The following guide should be added to the Autotracking In-Depth guide in the official guides, and the
sections on [POJOs][guides-pojos] and [Array][guides-arrays] should be removed.

### Tracked data structures

Generally, you should try to create classes with their tracked properties enumerated
and decorated with `@tracked`, instead of relying on dynamically created objects.
In some cases however, there could be a prohibitive number of possible properties,
or there could be no way to know them in advance.
In this case, you should use tracked versions of JavaScript's built-in

> Plain Old JavaScript Object (POJO), array or keyed collection (Maps, Sets, WeakMaps, WeakSets)

Reviewer 3: should we explain why certain structures are or aren't supported?
Like, I have some projects that make heavy use of [Uint32Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint32Array), I'm sure other folks have other stuff they'd like tracked versions of (or a way to optimally track them)

    this.items.push(item);
  }

  <template>
    {{#each this.items as |item|}}
      <div>
        {{item.productName}}
      </div>
    {{/each}}
  
    <button type="button" {{on "click" (fn this.addPants @pants)}}>
      Add Pants
    </button>
  </template>
}
```

## Drawbacks

> Why should we *not* do this? Please consider the impact on teaching Ember,
on the integration of this feature with other existing and planned features,
on the impact of the API churn on existing apps, etc.

> There are tradeoffs to choosing any path, please attempt to identify them here.

## Alternatives

- `tracked-built-ins` can be added only to the `app` blueprint and not to the `addon` blueprint.

- Do nothing and recommend using re-setting the property to track data structure changes.

  The tradeoffs of this approach described in "Motivation" section of this RFC.

- Don't add `tracked-built-ins` to the blueprint and only update Ember Guides.

  Ember.js is advertised as "batteries included" framework with complete out-of-the-box experience.
  Asking users to install additional dependencies to achieve relatively common problem contradicts
  with that paradigm.

## Unresolved questions

> 

Reviewer: Rereading this morning, I think we need another unresolved question here, which is:

```suggestion

- What, if any, substantive changes to the APIs presented by `tracked-built-ins` should we make?

```

For example (and I *strongly* suspect from previous discussions that other folks on Framework Core will agree with me on this), I think that while the decisions made around [what to support and what not to support from the actual built-in APIs](https://github.com/tracked-tools/tracked-built-ins/blob/9111be44d781448993c170fb68b59fb8a69667c2/README.md#usage) were perfectly reasonable, as we promote them to official status we probably want to revisit them. In particular:

- The `new TrackedArray()` and `new TrackedObject()` constructors differ from the `new Array()` and `new Object()` constructors; should we align them? For `Array`, it’s just a weird and confusing API, but aligning it with the built-in API is probably still fine: that way we just teach it as matching the built-in, however weird that is. For `new TrackedObject()`, though, we would have to mutate the original object, which seems weird.

- What about the omitted features? Can we make those “just work” by delegating to the class they are tracked versions of?

> - Should we introduce these at something like `@glimmer/tracking` instead?

Reviewer 3: This'd be my preference.

Also, providing an optional [`deepTracked`](https://github.com/NullVoxPopuli/ember-deep-tracked) for :musical_note: folks that just don't care :musical_note:

Reviewer: @NullVoxPopuli interesting note! I think it's probably fair to say that that definitely won't happen as part of *this* RFC. It may be worth adding, but needs its own RFC, because there are non-trivial design and *teaching* questions about what it looks like to include that. (My own take, not necessarily shared by anyone else on the Framework or TS core teams, is that using it is an anti-pattern outside of tests, and that there are generally much better ways of solving the needs people reach for that pattern for.)

Reviewer 3: (I agree lol)

whenever I've need deep tracking in practice, I've always found it easier to recursively use the cache primitives, so I can select what is tracked as I recurse

  ```js

>   import { tracked, TrackedMap, TrackedSet /* etc. */ } from '@glimmer/tracking';

Reviewer 4: Maybe a stupid question, but do we need the `TrackedMap`, `TrackedSet`, ... exports if we can do `tracked(Map)`, `tracked(Set)`, ...? To me, it seems nice to only have `import { tracked } from '@glimmer/tracking';`.

Reviewer 5: Hum, that fells less intuitive though. In practice I've always used the imports instead of wrapping.

Reviewer 3: the wrapping implies -- "well, what else can I wrap"?

Reviewer: This is really important point which I forgot to ask @SergeAstapov to address in drafting, @bertdeblock – thank you for raising it! I’ll make sure we cover it in our discussion (scheduled for Friday, but we’ll see if we get to it this week – I’m going to try!).

For my part (we haven't talked about it at Framework yet at all), this is actually a piece I *actively dislike* about the current thing and think we should remove. It makes `tracked` do two *very* different things, and that increases its complexity in non-trivial ways:

- **In terms of the “reasoning footprint” for what `tracked` is and does as a consumer:** from a pedagogical perspective, we would need to cover how it both works as a decorator *and* as a wrapper which produces non-decorated types, and indeed we would need to cover what this means (and why the decorator doesn't do *more* magic, which is a thing which comes up regularly already):

   ```js
   import { tracked } from '@glimmer/tracking';
   
   class Example {
      @tracked doublyTracked = tracked([]);
   }
   ```

- **In terms of the internal implementation details:** [supporting all of these different flows][impl] is not terrible, but it’s not trivial, and it also gets to just delegate to [the base decorator][base], so it’s more complicated than that shows.[^perf]

- **In terms of making sure the types do the right things:** On the types front, consider: if you want `tracked<T>(map: Map<T>): TrackedMap<T>` and `tracked<T extends object>(obj: T): TrackedObject<T>` and `tracked<T>(array: Array<T>): TrackedArray<T>` and so on to work, we have to provide overloads for all of them, and TS’s overload handling is order-sensitive, so we have to make sure that we don't mess *that* up, etc. ([Here][types] is how that is done today.)

All of that is *doable*; there are no critical blockers here. It’s more that it is complexity and overhead on all fronts, which needs to be well-motivated… and it’s not clear to *me* what the benefit is over importing and using the type you want directly—especially because all the types have constructors which allow them to be constructed from their “vanilla” versions.

Net, I think it's actually much better to be clear about what we’re actually doing.

(Again, that’s just my take; we’ll see what everyone else says when we discuss it!)

[impl]: https://github.com/tracked-tools/tracked-built-ins/blob/9111be44d781448993c170fb68b59fb8a69667c2/addon/-private/decorator.ts#L42-L86
[base]: https://github.com/glimmerjs/glimmer.js/blob/2696e52bae1ddaad160fc61cc1466fc47466cafc/packages/%40glimmer/tracking/src/tracked.ts
[types]: https://github.com/tracked-tools/tracked-built-ins/blob/9111be44d781448993c170fb68b59fb8a69667c2/addon/-private/decorator.ts#L13-L35

[^perf]: I don’t expect it would make a *huge* difference, but given how commonly `@tracked` is used and the fact that setting it up is part of booting the app, I would personally prefer to keep this stuff out of the default flow for performance-reasons, too.

Reviewer 5: Could also see it as a decorator

```ts
@tracked(Set)
ids: Set<number>
```

Reviewer 6: Given that `tracked` has to operate differently for each type of thing I don't think it is a great idea to overload its signature.

Reviewer 5: Yeah, probably right

Today, `@tracked` decorator may be used to track updates of the primitive values:

```js
import Component from '@glimmer/component';
import { tracked } from '@glimmer/tracking';
import { action } from '@ember/object';

export default class CounterComponent extends Component {
  @tracked count = 0;

  @action increment() {
    this.count = this.count + 1;
  }

  @action decrement() {
    this.count = this.count - 1;
  }
}
```

However, there are cases when there are unknown, dynamic, or potentially infinite numbers
of decoratable keys. In such cases we can use `Object`, `Map` or another data structure
that fits most in the particular use case. The initial pattern proposed for Ember Octane
in such scenarios was to re-set the tracked property:

```js
class TrackedMap {
  @tracked map = new Map();

  updateMapValue(key, value) {
    this.map.set(key, value);
    this.map = this.map;
  }
}
```

This pattern is problematic however, because it's not obvious what the
purpose of re-setting the property is. A developer unfamiliar with Ember may
mistakenly believe this was an error, and remove this statement, causing

> bugs. It's not a very google-able pattern either, which makes it more

Reviewer 7: Non essential but just thinking of real words :v: .

```suggestion
bugs. This pattern is not easy to search on the Internet, which makes it more
```

difficult to learn and to teach. Finally, it's fairly easy this way for the
state to get out of sync with the rest of the system if a developer forgets
to add the extra re-set after mutating the value.

> Over time, it's become clear that this pattern is in actuality an antipattern

Reviewer 7: Small typo I think... ?

```suggestion
Over time, it's become clear that this pattern is actually an anti-pattern
```

  TrackedWeakSet,
} from 'tracked-built-ins';
```

For example, we could make a scoreboard component that keeps score for an
arbitrary number of players, and keep track of the score for each player using
tracked version of JavaScript [Map][global-objects-map].

**Note:** In this and the following examples, I am using `<template>` from
[RFC #779][rfc-0779].

```js
// app/components/scoreboard.js
import Component from '@glimmer/component';
import { action } from '@ember/object';
import { TrackedMap } from 'tracked-built-ins';

const PLAYERS = [
  { name: 'Zoey' },
  { name: 'Tomster' },
];

export default class Scoreboard extends Component {
  // Create a map from player -> score, with each player's score starting at 0
  scores = new TrackedMap(PLAYERS.map(p => [p, 0]));

  @action
  incrementScore(player) {
    let currentScore = this.scores.get(player);

    this.scores.set(player, currentScore + 1);
  }

  <template>
    {{#each-in this.scores |player score|}}
      <div>
        {{player.name}}: {{score}}
    
        <button type="button" {{on "click" (fn this.incrementScore player)}}>
          +
        </button>
      </div>
    {{/each-in}}
  </template>
}
```

> Lets lake a look into another example involving a shopping cart, where we keep track

Reviewer 7: Small typo.

```suggestion
Let's take a look into another example involving a shopping cart, where we keep track
```

in order to avoid its mutation which may lead to issues.

The tracked data structures provided by `tracked-built-ins` addon are *not* deeply tracked.
This can be demonstrated in such example:

```ts
import Component from '@glimmer/component';
import { action } from '@ember/object';
import { TrackedObject } from 'tracked-built-ins';

export default class Scoreboard extends Component {
  obj = new TrackedObject({
    foo: {
      bar: 'baz'
    },
    bar: 'baz'
  });

  @action
  myAction() {
    this.obj.bar = 'barBaz'; // is autotracked and triggers re-render.
    this.obj.foo.bar = 'barBaz'; // is *not* autotracked and *does not* trigger re-render.
  }

  <template>
    bar: {{this.obj.bar}}
    foo.bar: {{this.obj.foo.bar}}

    <button type="button" {{on "click" this.myAction}}>press me</button>
  </template>
}
```

There several reasons for that limitation:
 - the correct semantics for a framework-level hook there aren't obvious.
 - it is possible to build variants on it fairly cheaply in user-land given `tracked()` as updated here.

## How we teach this

The following guide should be added to the Autotracking In-Depth guide in the official guides, and the

> sections on [POJOs][guides-pojos] and [Array][guides-arrays] should be removed.

Reviewer 8: I might be wrong, but the 4.11 docs do not seem to have been updated to reflect the addition of the `tracked-built-ins` addon and how to use it like this RFC suggested

Reviewer 4: I think at this point, this RFC could/should be moved to the "Released" stage.
Making sure the docs are updated are definitely a requirement for getting it to the "Recommended" stage if I'm not mistaken.
